test/rebar3_typer_prv_SUITE.erl

```diff
@@ -0,0 +1,271 @@
+     recursive,
+     good_modes,
+     colliding_modes,
+     show_success_typings,
+     no_spec,
+     edoc,
+     plt,
+     typespec_files,
+     unrecognized_opt,
+     format_error].
+
+init_per_testcase(_, Config) ->
+    Self = self(),
+    meck:new(rebar3_mini_typer),
+    meck:expect(rebar3_mini_typer,
+                run,
+                fun(Opts) ->
+                   Self ! #{opts => Opts},
+                   ok
+                end),
+    Config.
+
+end_per_testcase(_, Config) ->
+    meck:unload(rebar3_mini_typer),
+    Config.
+
+%% @doc Just try to run typer without options
+no_options(_Config) ->
+    {ok, State} =
+        rebar3_typer:init(
+            rebar_state:new()),
+
+    ct:comment("Simply running typer without any parameter should use only default values"),
+    [{mode, show}] = get_opts(State),
+
+    {comment, ""}.
+
+%% @doc --recursive / recursive
+recursive(_Config) ->
+    {ok, State} =
+        rebar3_typer:init(
+            rebar_state:new()),
+
+    ct:comment("files_r is correctly picked up from rebar.config"),
+    Files = ["src/", "test/"],
+    State1 = rebar_state:set(State, typer, [{recursive, Files}]),
+    {files_r, Files} = lists:keyfind(files_r, 1, get_opts(State1)),
+
+    ct:comment("--recursive takes precedence"),
+    State2 = rebar_state:command_parsed_args(State1, {[{recursive, "lib/,src/"}], []}),
+    {files_r, ["lib/", "src/"]} = lists:keyfind(files_r, 1, get_opts(State2)),
+    {comment, ""}.
+
+%% @doc --show|show_exported|annotate|annotate_inc_files / mode
+good_modes(_Config) ->
+    {ok, State} =
+        rebar3_typer:init(
+            rebar_state:new()),
+
+    ct:comment("mode is correctly picked up from rebar.config"),
+    State1 = rebar_state:set(State, typer, [{mode, annotate}]),
+    {mode, annotate} = lists:keyfind(mode, 1, get_opts(State1)),
+
+    State2 = rebar_state:set(State1, typer, [{mode, show_exported}]),
+    {mode, show_exported} = lists:keyfind(mode, 1, get_opts(State2)),
+
+    ct:comment("--show takes precedence"),
+    State3 = rebar_state:command_parsed_args(State2, {[{show, true}], []}),
+    {mode, show} = lists:keyfind(mode, 1, get_opts(State3)),
+
+    ct:comment("--show=false uses what's in rebar.config"),
+    State4 = rebar_state:command_parsed_args(State3, {[{show, false}], []}),
+    {mode, show_exported} = lists:keyfind(mode, 1, get_opts(State4)),
+
+    ct:comment("--annotate works"),
+    State5 = rebar_state:command_parsed_args(State4, {[{annotate, true}], []}),
+    {mode, annotate} = lists:keyfind(mode, 1, get_opts(State5)),
+
+    ct:comment("--show_exported works"),
+    State6 = rebar_state:command_parsed_args(State5, {[{show_exported, true}], []}),
+    {mode, show_exported} = lists:keyfind(mode, 1, get_opts(State6)),
+
+    ct:comment("--annotate-inc-files works"),
+    State7 = rebar_state:command_parsed_args(State6, {[{annotate_inc_files, true}], []}),
+    {mode, annotate_inc_files} = lists:keyfind(mode, 1, get_opts(State7)),
```

Comment:
Doesn't using the previous state each time make it behave like having lots of modes all lined up on the CLI? Why not have them testing against State2 if you're comparing to having things in the config?

Reply:
Great point. It was not a problem, since `command_parsed_args` just replaced everything that came before it… but it's worth maintaining proper semantics in our tests. 💪🏻